src/keys.rs

```diff
@@ -0,0 +1,112 @@
+//! Key structures for Orchard.
+
+use crate::address::Address;
+
+/// A spending key, from which all key material is derived.
+///
+/// TODO: In Sapling we never actually used this, instead deriving everything via ZIP 32,
+/// so that we could maintain Bitcoin-like HD keys with properties like non-hardened
+/// derivation. If we decide that we don't actually require non-hardened derivation, then
+/// we could greatly simplify the HD structure and use this struct directly.
+#[derive(Debug)]
+pub struct SpendingKey;
+
+#[derive(Debug)]
+pub(crate) struct SpendAuthorizingKey;
+
+impl From<&SpendingKey> for SpendAuthorizingKey {
+    fn from(_: &SpendingKey) -> Self {
+        todo!()
+    }
+}
+
+/// TODO: This is its protocol spec name for Sapling, but I'd prefer a different name.
+#[derive(Debug)]
+pub(crate) struct AuthorizingKey;
+
+impl From<&SpendAuthorizingKey> for AuthorizingKey {
+    fn from(_: &SpendAuthorizingKey) -> Self {
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NullifierDerivingKey;
+
+impl From<&SpendingKey> for NullifierDerivingKey {
+    fn from(_: &SpendingKey) -> Self {
+        todo!()
+    }
+}
+
+/// A key that provides the capability to recover outgoing transaction information from
+/// the block chain.
+#[derive(Debug)]
+pub struct OutgoingViewingKey;
+
+impl From<&SpendingKey> for OutgoingViewingKey {
+    fn from(_: &SpendingKey) -> Self {
+        todo!()
+    }
+}
+
+/// A key that provides the capability to view incoming and outgoing transactions.
+///
+/// This key is useful in situations where you only need the capability to detect inbound
+/// payments, such as merchant terminals.
+///
+/// This key is not suitable for use in a wallet, as it cannot maintain accurate balance.
+/// You should use a [`FullViewingKey`] instead.
+///
```

Comment:
```suggestion
```

Reply:
Why remove this?

Reply:
This is repeated on line 90

Reply:
This was on FVK, not IVK. Copy-paste error on my part.